### PR TITLE
Fleet navigation ROS 2 package and assets

### PR DESCRIPTION
**Pull request by pijaro**
`Ros2FleetRobotTemplate` update.

A fleet navigation ROS 2 package was added to the template. It is based on nav2_bringup package.

Navigation with AMCL for multiple robots.

Assumes that the namespace spawning is implemented (https://github.com/o3de/o3de-extras/pull/444).

Zip archive will be added after the review:

- [ ] Zip archive updated.

### CODE REVIEW

File: Templates/Ros2FleetRobotTemplate/README.md [hunk 1]
```diff
@@ -15,30 +15,109 @@ If your simulation does not match this and you would like to start with a simple
 Please follow the instructions in [ROS 2 Gem documentation](https://development--o3deorg.netlify.app/docs/user-guide/interactivity/robotics/project-configuration/)
 to install all required dependencies and create your project with a template (make sure to use chose this template during the process).
 
-## Spawning robots
+## Fleet navigation
 
-The level contains spawn points configured to easily add more Proteus robots through ROS 2 calls.
```
Comment (adamdbrw) [inline Templates/Ros2FleetRobotTemplate/README.md:20]: How about to leave this part as "Simple spawning" and structure in a way that users know to follow either simple spawning or running the fleet example?

File: Templates/Ros2FleetRobotTemplate/Template/Examples/ros2_ws/src/o3de_fleet_nav/params/nav2_params.yaml [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+    global_frame: map
+    robot_base_frame: base_link
+    odom_topic: odom
+    bt_loop_duration: 10
+    default_server_timeout: 20
+    # 'default_nav_through_poses_bt_xml' and 'default_nav_to_pose_bt_xml' are use defaults:
+    # nav2_bt_navigator/navigate_to_pose_w_replanning_and_recovery.xml
+    # nav2_bt_navigator/navigate_through_poses_w_replanning_and_recovery.xml
+    # They can be set here or via a RewrittenYaml remap from a parent launch file to Nav2.
+    plugin_lib_names:
```
Comment (adamdbrw) [inline Templates/Ros2FleetRobotTemplate/Template/Examples/ros2_ws/src/o3de_fleet_nav/params/nav2_params.yaml:56]: these files seem to be based on Humble version. Did you test with Iron? See #451

File: Templates/Ros2FleetRobotTemplate/README.md [hunk 3]
```diff
@@ -15,30 +15,109 @@ If your simulation does not match this and you would like to start with a simple
...
+- name (a type of the robot to spawn), 
+- namespace (must be unique per spawned robot),
+- spawning position (spawn position is also used as an AMCL initial estimation). 
+
+In this example, only the `proteus` robot is supported.
+
+You can modify contents of this file to add/remove robots or change their initial positions.
+
+> Notice: You have to rebuild the ROS 2 workspace for changes to update.
+
```
Comment (adamdbrw) [inline Templates/Ros2FleetRobotTemplate/README.md:62]: One of the most important use cases would be importing own robot and setting it with navigation stack. I believe instructions should provide helpful tips on what needs to be adjusted in navigation stack parameters in order for the new robot to work. Of course this has to be on a general level, and sending users to external documentation resources.

File: Templates/Ros2FleetRobotTemplate/README.md (hunk 1 above)
Comment (pijaro) [reply Templates/Ros2FleetRobotTemplate/README.md:20]: I've put the "simple spawning" section back, with a minor notice, that for the fleet nav, the scene must be empty (robots are spawning automatically).

File: Templates/Ros2FleetRobotTemplate/Template/Examples/ros2_ws/src/o3de_fleet_nav/params/nav2_params.yaml (hunk 2 above)
Comment (pijaro) [reply Templates/Ros2FleetRobotTemplate/Template/Examples/ros2_ws/src/o3de_fleet_nav/params/nav2_params.yaml:56]: Added support for humble/iron parameters.

Tested on iron and humble.

File: Templates/Ros2FleetRobotTemplate/README.md (hunk 3 above)
Comment (pijaro) [reply Templates/Ros2FleetRobotTemplate/README.md:78]: Added a section about preparing robot :+1:

File: Templates/Ros2FleetRobotTemplate/template.json [hunk 4]
```diff
@@ -266,7 +270,100 @@
...
+        {
+            "file": "Examples/ros2_ws/src/o3de_fleet_nav/config/fleet_config.yaml",
+            "isTemplated": false
+        },
+        {
+            "file": "Examples/ros2_ws/src/o3de_fleet_nav/launch/o3de_rviz_launch.py",
+            "isTemplated": false
+        },
+        {
+            "file": "Examples/ros2_ws/src/o3de_fleet_nav/launch/o3de_slam_launch.py",
```
Comment (lewandowskikacper) [inline Templates/Ros2FleetRobotTemplate/template.json:315]: There is no `o3de_slam_launch.py` file in the template.
This causes project creation to fail.

File: Templates/Ros2FleetRobotTemplate/README.md [hunk 5]
```diff
@@ -28,17 +29,135 @@ ros2 service call /spawn_entity gazebo_msgs/srv/SpawnEntity '{name: 'proteus', x
...
+### Building
+
+- Source ROS 2:
+```
+. /opt/ros/humble/setup.bash
+```
+
+- Go to the ROS 2 workspace:
+```
+cd Example/ros2_ws
```
Comment (lewandowskikacper) [inline Templates/Ros2FleetRobotTemplate/README.md:127]: This directory is called `Examples`, not `Example`

File: Templates/Ros2FleetRobotTemplate/README.md [hunk 6]
```diff
@@ -28,17 +29,135 @@ ros2 service call /spawn_entity gazebo_msgs/srv/SpawnEntity '{name: 'proteus', x
...
+- Run the `Warehouse` level in O3DE editor.
+
+- Source ROS 2 and the `o3de_fleet_nav` workspace:
+```
+. /opt/ros/humble/setup.bash
+```
+
+- Source the `o3de_fleet_nav` workspace:
+```
+cd Example/ros2_ws
```
Comment (lewandowskikacper) [inline Templates/Ros2FleetRobotTemplate/README.md:152]: This directory is called `Examples`, not `Example`

**Review by lewandowskikacper — COMMENTED** [review]
Params are not copied correctly during project creation from this template (`humble` and `iron` subdirectories are not copied). This causes a crash when launching `o3de_fleet_nav_launch.py` as in the instruction.

```
[INFO] [launch]: Default logging verbosity is set to INFO
ReplaceString substitution error:  [Errno 2] No such file or directory: '/home/kacper/projects/FleetNavigation/Examples/ros2_ws/install/o3de_fleet_nav/share/o3de_fleet_nav/params/humble/nav2_multirobot_params.yaml'
[ERROR] [launch]: Caught exception in launch (see debug for traceback): local variable 'input_file' referenced before assignment
```

Copying the files from ~/o3de-extras/Templates/Ros2FleetRobotTemplate/Template/Examples/ros2_ws/src/o3de_fleet_nav/params/ to ~/projects/FleetNavigation/Examples/ros2_ws/src/o3de_fleet_nav/params/ fixes that. 

Possibly, those files should be included in `template.json`?

**Comment by lewandowskikacper** [comment]
I created a project from this template and run it (on humble).
Apart from the problems listed above, the project and example run as expected.

Some comment about map creation could be useful (how to update it, when the warehouse is modified).

**Comment by adamdbrw** [comment]
@lewandowskikacper @pijaro are all issues resolved?

File: Templates/Ros2FleetRobotTemplate/README.md [hunk 7]
```diff
@@ -28,17 +29,135 @@ ros2 service call /spawn_entity gazebo_msgs/srv/SpawnEntity '{name: 'proteus', x
...
+- `/robot1/odom`
+- `/robot1/base_link`
+- `/robot1/lidar`
 
 To understand more about transformations, see ROS 2 navigation [documentation](https://navigation.ros.org/setup_guides/transformation/setup_transforms.html).
+
+## Using your robots in the simulation
+
+You can also use your robots in the simulation. To do so, you need to:
+- [import robot](https://docs.o3de.org/docs/user-guide/interactivity/robotics/importing-robot/) from existing URDF file,
```
Comment (adamdbrw) [inline Templates/Ros2FleetRobotTemplate/README.md:105]:
```suggestion
- [import a robot](https://docs.o3de.org/docs/user-guide/interactivity/robotics/importing-robot/) from existing URDF file,
```

File: Templates/Ros2FleetRobotTemplate/README.md [hunk 8]
```diff
@@ -28,17 +29,135 @@ ros2 service call /spawn_entity gazebo_msgs/srv/SpawnEntity '{name: 'proteus', x
...
+  - or create a robot from scratch in a O3DE editor using ROS 2 gem components (see [Frames](https://docs.o3de.org/docs/user-guide/interactivity/robotics/concepts-and-components-overview/#frames)),
```
Comment (adamdbrw) [inline Templates/Ros2FleetRobotTemplate/README.md:106]:
```suggestion
  - or create a robot from scratch in a O3DE editor using ROS 2 gem components (see [Frames](https://docs.o3de.org/docs/user-guide/interactivity/robotics/concepts-and-components-overview)),
```

File: Templates/Ros2FleetRobotTemplate/README.md [hunk 9]
```diff
@@ -28,17 +29,135 @@ ros2 service call /spawn_entity gazebo_msgs/srv/SpawnEntity '{name: 'proteus', x
...
+cd Examples/ros2_ws
+. ./install/setup.bash
+```
+
+- Run the fleet example:
+```
+ros2 launch o3de_fleet_nav o3de_fleet_nav_launch.py
+```
+
+Few Rviz2 windows should appear. You can use the "Nav2 Goal" button to send goal to the robot.
```
Comment (adamdbrw) [inline Templates/Ros2FleetRobotTemplate/README.md:161]:
```suggestion
Few RViz2 windows should appear. You can use the "Nav2 Goal" button to send goal to the robot.
```

File: Templates/Ros2FleetRobotTemplate/Template/Examples/ros2_ws/src/o3de_fleet_nav/NOTICE [hunk 10]
```diff
@@ -0,0 +1 @@
+This package contains a modified code from "nav2_bringup" package (https://github.com/ros-planning/navigation2).
```
Comment (adamdbrw) [inline Templates/Ros2FleetRobotTemplate/Template/Examples/ros2_ws/src/o3de_fleet_nav/NOTICE:1]:
```suggestion
This package contains a modified code from "nav2_bringup" package (https://github.com/ros-planning/navigation2).

```

File: Templates/Ros2FleetRobotTemplate/Template/Examples/ros2_ws/src/o3de_fleet_nav/launch/o3de_fleet_nav_launch.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
+# Copyright (c) 2018 Intel Corporation
```
Comment (adamdbrw) [inline Templates/Ros2FleetRobotTemplate/Template/Examples/ros2_ws/src/o3de_fleet_nav/launch/o3de_fleet_nav_launch.py:1]:
```suggestion
```

File: Templates/Ros2FleetRobotTemplate/Template/Examples/ros2_ws/src/o3de_fleet_nav/launch/o3de_rviz_launch.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
+# Copyright (c) 2018 Intel Corporation
```
Comment (adamdbrw) [inline Templates/Ros2FleetRobotTemplate/Template/Examples/ros2_ws/src/o3de_fleet_nav/launch/o3de_rviz_launch.py:1]:
```suggestion
```

File: Templates/Ros2FleetRobotTemplate/Template/Examples/ros2_ws/src/o3de_fleet_nav/launch/o3de_nav_launch.py [hunk 13]
```diff
@@ -0,0 +1,256 @@
+# Copyright (c) 2018 Intel Corporation
```
Comment (adamdbrw) [inline Templates/Ros2FleetRobotTemplate/Template/Examples/ros2_ws/src/o3de_fleet_nav/launch/o3de_nav_launch.py:1]:
```suggestion
```

**Comment by pijaro** [comment]
I was thinking about adding some launch file for building up a map. On the other hand, I think that I can add it in separate PR.

I will take into

File: Templates/Ros2FleetRobotTemplate/Template/Examples/ros2_ws/src/o3de_fleet_nav/launch/o3de_nav_launch.py [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+
+    declare_use_namespace_cmd = DeclareLaunchArgument(
+        'use_namespace',
+        default_value='false',
+        description='Whether to apply a namespace to the navigation stack')
+
+    declare_map_yaml_cmd = DeclareLaunchArgument(
+        'map',
+        default_value=os.path.join(
+            o3de_fleet_nav_dir, 'maps', 'map.yaml'),
```
Comment (lewandowskikacper) [inline Templates/Ros2FleetRobotTemplate/Template/Examples/ros2_ws/src/o3de_fleet_nav/launch/o3de_nav_launch.py:58]: There is no `map.yaml` in `maps`, so this fails if left on default.

```suggestion
            o3de_fleet_nav_dir, 'maps', 'map_warehouse.yaml'),
```

**Comment by pijaro** [comment]
Review applied.

As for the map creation, we can go with a new issue and work on slam/map building example :+1:

**Review by adamdbrw — APPROVED** [review]

**Review by michalpelka — APPROVED** [review]
Tested and reviewed. Please merge #444 ASAP.

**Comment by michalpelka** [comment]
rebased

**Pull request merged**